NickvisionCavalier.GNOME/Views/PreferencesDialog.cs

```diff
@@ -288,18 +306,150 @@ private PreferencesDialog(Gtk.Builder builder, PreferencesViewController control
+                var color = (Color)Marshal.PtrToStructure(colorPtr, typeof(Color));
+                if (color.Alpha <= 0.0001f)
+                {
+                    color.Red = 0;
+                    color.Green = 0;
+                    color.Blue = 0;
+                }
+                _controller.AddColor(ColorType.Foreground, $"#{((int)(color.Alpha * 255)).ToString("x2")}{((int)(color.Red * 255)).ToString("x2")}{((int)(color.Green * 255)).ToString("x2")}{((int)(color.Blue * 255)).ToString("x2")}");
+                UpdateColorsGrid();
+            }
+        };
+        _bgColorDialogCallback = (source, res, data) =>
+        {
+            var colorPtr = gtk_color_dialog_choose_rgba_finish(_colorDialog, res, IntPtr.Zero);
+            if (colorPtr != IntPtr.Zero)
+            {
+                var color = (Color)Marshal.PtrToStructure(colorPtr, typeof(Color));
+                if (color.Alpha <= 0.0001f)
+                {
+                    color.Red = 0;
+                    color.Green = 0;
+                    color.Blue = 0;
+                }
+                _controller.AddColor(ColorType.Background, $"#{((int)(color.Alpha * 255)).ToString("x2")}{((int)(color.Red * 255)).ToString("x2")}{((int)(color.Green * 255)).ToString("x2")}{((int)(color.Blue * 255)).ToString("x2")}");
+                UpdateColorsGrid();
+            }
+        };
+        _addFgColorButton.OnClicked += (sender, e) => AddColor(ColorType.Foreground);
+        _addBgColorButton.OnClicked += (sender, e) => AddColor(ColorType.Background);
+        UpdateColorProfiles();
     }
 
     /// <summary>
     /// Constructs a PreferencesDialog
     /// </summary>
     /// <param name="controller">PreferencesViewController</param>
-    /// <param name="application">Adw.Application</param>
-    /// <param name="parent">Gtk.Window</param>
-    public PreferencesDialog(PreferencesViewController controller, Adw.Application application, Gtk.Window parent) : this(Builder.FromFile("preferences_dialog.ui"), controller, application, parent)
+    public PreferencesDialog(PreferencesViewController controller) : this(Builder.FromFile("preferences_dialog.ui"), controller)
+    {
+    }
+
+    private void UpdateColorProfiles()
+    {
+        _profilesList.SelectRow(null);
+        while (_profilesList.GetRowAtIndex(0) != null)
+        {
+            _profilesList.Remove(_profilesList.GetRowAtIndex(0));
+        }
+        for (var i = 0; i < _controller.ColorProfiles.Count; i++)
+        {
+            var profileBox = new ProfileBox(_controller.ColorProfiles[i].Name, i);
+            _profilesList.Append(profileBox);
+            profileBox.OnDelete += OnDeleteProfile;
+        }
+        _profilesList.SelectRow(_profilesList.GetRowAtIndex(_controller.ActiveProfile));
+    }
+
+    private void OnAddProfile(object sender, EventArgs e)
+    {
+        var dialog = new AddProfileDialog(this, _controller.AppInfo.ID);
+        dialog.OnDialogClosed += (sender, response) =>
```

Comment:
See above comments...this can just be `dialog.OnReponse` and then you can use `dialog.Response` to get the actual response...